File: bc2026/src/bestsofar/Utils/Symmetry.cpp
#include "Utils/Symmetry.hpp"

#include <iostream>

#include "Utils/Comms.hpp"
#include "Utils/Globals.hpp"
#include "Utils/Vision.hpp"

namespace Utils {

int Symmetry::symm = 0;  // 0: possible, 1: impossible

bool Symmetry::isSymmetryPossible(int s) {
    return (symm & s) != 0;
}

MapLocation Symmetry::symmetricallyOpposite(MapLocation loc, int s) {
    if (s == vertical) return MapLocation(mapWidth - 1 - loc.x, loc.y);
    if (s == horizontal) return MapLocation(loc.x, mapHeight - 1 - loc.y);
    if (s == rotational) return MapLocation(mapWidth - 1 - loc.x, mapHeight - 1 - loc.y);
    return MapLocation::NONE;
}

vector<int> Symmetry::getPossibleSymmetries() {
    if (symm == 0) return {1, 2, 4};
    if (symm == 1) return {2, 4};
    if (symm == 2) return {1, 4};
    if (symm == 4) return {2, 4};
    if (symm == 3) return {4};
    if (symm == 5) return {2};
    if (symm == 6) return {1};
    rc.setIndicatorString("what the hecky? no symmetries!");
    std::cout << "what the hecky? no symmetries!" << '\n';
    return {};  // Java: return null (callers then NPE - see notes)
}

void Symmetry::updateNearbySymmetry() {
    vector<int> possibleSymmetries = getPossibleSymmetries();

    // special case king to not bytecode exceed cuz its vision is bigger
    if (isRatKingType(rc.getType())) {
        for (const MapLocation& loc : rc.getAllLocationsWithinRadiusSquared(
                 rc.getLocation(), Vision::visionConeRadiusSquared(rc.getType()))) {
            MapInfo map = rc.senseMapInfo(loc);

            for (const int& s : possibleSymmetries) {
                MapLocation opp = symmetricallyOpposite(loc, s);

                if (Vision::hasSeenLocation(opp)) {
                    if (map.hasCheeseMine() != Vision::isMine(opp)) symm |= s;
                    if (map.isWall() != Vision::isWall(opp)) symm |= s;
                }
            }
        }
        return;
    }

    // baby rat
    for (const MapInfo& map : surroundings) {
        MapLocation loc = map.getMapLocation();

        bool mine = map.hasCheeseMine();
        bool wall = map.isWall();

        if (!mine && !wall) continue;  // save bytecode by only checking if the current cell is special

        for (const int& s : possibleSymmetries) {
            MapLocation opp = symmetricallyOpposite(loc, s);

            if (Vision::hasSeenLocation(opp)) {
                if (mine != Vision::isMine(opp)) symm |= s;
                if (wall != Vision::isWall(opp)) symm |= s;
            }
        }
    }
}

void Symmetry::updateSymmetryFromGlobal() {
    int globalSymm = Comms::getSymmetry();
    symm |= globalSymm;
}

void Symmetry::commSymmetryToNearby() {
    int globalSymm = Comms::getSymmetry();

    // only if is within hearing distance of king
    if (rc.getLocation().isWithinDistanceSquared(kingPos[0], 16)) {
        if (symm != globalSymm) {
            rc.squeak(symm);
            hasSqueaked = true;
            std::cout << "squeaked new symmetry to king: " + std::to_string(symm) << '\n';
        }
    }
    // squeak every round because we're doing it anyways
    else if (!hasSqueaked && Symmetry::symm != globalSymm) {
        rc.squeak(Symmetry::symm);
        hasSqueaked = true;
    }
}

void Symmetry::hearNearbySymmetrySqueaks() {
    for (const Message& message : newestSqueaks) {
        if (0 <= message.getBytes() && message.getBytes() < 8) {
            symm |= message.getBytes();
        }
    }
}

}  // namespace Utils

File: py/bc2026/src/bestsofar/Utils/Symmetry.py
# Port of bestsofar/Utils/Symmetry.java. See py/CONVENTIONS.md.
from api import *
from Utils.Comms import Comms
from Utils.Globals import Globals
from Utils.Vision import Vision


class Symmetry(Globals):

    vertical = 1
    horizontal = 2
    rotational = 4

    symm = 0  # 0: possible, 1: impossible

    @staticmethod
    def isSymmetryPossible(s):
        return (Symmetry.symm & s) != 0

    @staticmethod
    def symmetricallyOpposite(loc, s):
        if s == Symmetry.vertical:
            return MapLocation(Globals.mapWidth - 1 - loc.x, loc.y)
        if s == Symmetry.horizontal:
            return MapLocation(loc.x, Globals.mapHeight - 1 - loc.y)
        if s == Symmetry.rotational:
            return MapLocation(Globals.mapWidth - 1 - loc.x, Globals.mapHeight - 1 - loc.y)
        return None

    @staticmethod
    def getPossibleSymmetries():
        if Symmetry.symm == 0:
            return [1, 2, 4]
        if Symmetry.symm == 1:
            return [2, 4]
        if Symmetry.symm == 2:
            return [1, 4]
        if Symmetry.symm == 4:
            return [2, 4]
        if Symmetry.symm == 3:
            return [4]
        if Symmetry.symm == 5:
            return [2]
        if Symmetry.symm == 6:
            return [1]
        Globals.rc.setIndicatorString("what the hecky? no symmetries!")
        print("what the hecky? no symmetries!")
        return None

    @staticmethod
    def updateNearbySymmetry():
        possibleSymmetries = Symmetry.getPossibleSymmetries()

        # special case king to not bytecode exceed cuz its vision is bigger
        if isRatKingType(Globals.rc.getType()):  # Java: rc.getType().isRatKingType()
            # Java: rc.getType().getVisionRadiusSquared()
            for loc in Globals.rc.getAllLocationsWithinRadiusSquared(Globals.rc.getLocation(), Vision.visionConeRadiusSquared(Globals.rc.getType())):
                map = Globals.rc.senseMapInfo(loc)

                for s in possibleSymmetries:
                    opp = Symmetry.symmetricallyOpposite(loc, s)

                    if Vision.hasSeenLocation(opp):
                        if map.hasCheeseMine() != Vision.isMine(opp):
                            Symmetry.symm |= s
                        if map.isWall() != Vision.isWall(opp):
                            Symmetry.symm |= s
            return

        # baby rat
        for map in Globals.surroundings:
            loc = map.getMapLocation()

            mine = map.hasCheeseMine()
            wall = map.isWall()

            if not mine and not wall:
                continue  # save bytecode by only checking if the current cell is special

            for s in possibleSymmetries:
                opp = Symmetry.symmetricallyOpposite(loc, s)

                if Vision.hasSeenLocation(opp):
                    if mine != Vision.isMine(opp):
                        Symmetry.symm |= s
                    if wall != Vision.isWall(opp):
                        Symmetry.symm |= s

    @staticmethod
    def updateSymmetryFromGlobal():
        globalSymm = Comms.getSymmetry()
        Symmetry.symm |= globalSymm

    @staticmethod
    def commSymmetryToNearby():
        globalSymm = Comms.getSymmetry()

        # only if is within hearing distance of king
        if Globals.rc.getLocation().isWithinDistanceSquared(Globals.kingPos[0], 16):
            if Symmetry.symm != globalSymm:
                Globals.rc.squeak(Symmetry.symm)
                Globals.hasSqueaked = True
                print("squeaked new symmetry to king: " + str(Symmetry.symm))
        # squeak every round because we're doing it anyways
        elif not Globals.hasSqueaked and Symmetry.symm != globalSymm:
            Globals.rc.squeak(Symmetry.symm)
            Globals.hasSqueaked = True

    @staticmethod
    def hearNearbySymmetrySqueaks():
        for message in Globals.newestSqueaks:
            if 0 <= message.getBytes() and message.getBytes() < 8:
                Symmetry.symm |= message.getBytes()
